src/main/AllocationTable.cpp

```diff
@@ -0,0 +1,334 @@
+    erase(pointer, bucket->size);
+    // Add allocation to free list.
+    add_free_block(pointer, bucket->size);
+  }
+  return true;
+}
+
+void AllocationTable::ApplyToAllBlocks(std::function<void(void*)> fun) {
+  for (int i = 0; i < num_buckets; ++i) {
+    HashBucket& bucket = buckets[i];
+    if (bucket.data != EMPTY_KEY && bucket.data != DELETED_KEY) {
+      if (bucket.data & 1) {
+        uintptr_t pointer = bucket.DataPointer();
+        if (bucket.data & 2) {
+          // This is a large superblock.
+          Superblock* superblock = reinterpret_cast<Superblock*>(pointer);
+          uintptr_t block = pointer + SUPERBLOCK_HEADER_SIZE;
+          uintptr_t block_end = pointer + MEDIUM_SUPERBLOCK_SIZE;
+          unsigned block_size = 1 << superblock->block_size;
+          unsigned superblock_size =
+              (SUPERBLOCK_SIZE - SUPERBLOCK_HEADER_SIZE)
+              / superblock->block_size;
+          unsigned bitset_entries =
+              ((1 << (MEDIUM_SUPERBLOCK_SIZE_LOG2 - superblock->block_size))
+                  + 63) / 64;
+          for (int i = 0; i < bitset_entries; ++i) {
+            if (superblock->free[i] != ~0ull) {
+              for (int index = 0; index < 64; ++index) {
+                if (!(superblock->free[i] & (u64{1} << index))) {
+#ifdef ALLOCATION_CHECK
+                  if (!lookup_in_allocation_map(reinterpret_cast<void*>(
+                      block))) {
+                    allocerr("apply to all blocks iterating over "
+                        "non-alloc'd pointer");
+                  }
+#endif
+#ifdef HAVE_ADDRESS_SANITIZER
+                  fun(offset_pointer(reinterpret_cast<void*>(block),
+                      REDZONE_SIZE));
+#else
+                  fun(reinterpret_cast<void*>(block));
+#endif
+                }
+                block += block_size;
+              }
+            } else {
+              block += block_size * 64;
+            }
+          }
+        } else {
+#ifdef HAVE_ADDRESS_SANITIZER
+          fun(offset_pointer(reinterpret_cast<void*>(pointer), REDZONE_SIZE));
+#else
+          fun(reinterpret_cast<void*>(pointer));
+#endif
+        }
+      }
+    }
+  }
+}
+
+void AllocationTable::PrintTable() {
+  printf(">>>>> SPARSE TABLE\n");
+  int size = 0;
+  for (int i = 0; i < num_buckets; i += 16) {
+    int count = 0;
+    for (int j = 0; j < 16 && i + j < num_buckets; ++j) {
+      HashBucket& bucket = buckets[i + j];
+      if (bucket.data != EMPTY_KEY && bucket.data != DELETED_KEY) {
+        count += 1;
+      }
+    }
+    if (count) {
+      printf("%X", count);
+    } else {
+      printf(".");
+    }
+    size += count;
+  }
+  printf("\n");
+  printf("table size: %d\n", size);
+}
+
+bool AllocationTable::try_insert(HashBucket* table, uintptr_t pointer, size_t size) {
+  uintptr_t key = pointer >> LOW_BITS;
+  uintptr_t key_end = (pointer + (1 << size) + ((1 << LOW_BITS) - 1))
+      >> LOW_BITS;
+  bool first = true;
+  do {
```

Comment:
It gets awkward with a for loop, because this is one of those rare occasions when the loop should always take one iteration even if the first key is equal to the end key. If the first key is not equal to the end key then the end key should not be inserted.